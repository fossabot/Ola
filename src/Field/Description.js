import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const FieldDescription = ({ error, children }) => {
  return (
    <p className={cx({'ola_field-error': error, 'ola_field-description': !error })}>
      { children }
    </p>
  )
}

FieldDescription.defaultProps = {
  error: false
}

FieldDescription.propTypes = {
  /** Error variant (automatic if Field is parent) */
  error: PT.bool,
  /** Childen input node */
  children: PT.oneOfType([
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired
}

export default FieldDescription
