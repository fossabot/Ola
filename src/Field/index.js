import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const Field = ({ id, label, hint, error, description, customDescription, disabled, children }) => {
  return (
    <div className={cx('field', {'is-invalid': error}, {'is-disabled': disabled})}>
      <label htmlFor={id} className="field-label">
        { label }
        {hint && <span className="field-hint">{ hint }</span>}
      </label>
      <div className="field-input">
        {React.cloneElement(children, { id: id, error, disabled })}
      </div>
      {
        customDescription ?
          React.cloneElement(customDescription, { error }) :
          ( <p className={ cx({ 'field-error': error, 'field-description': !error }) }>{description}</p> )
      }

    </div>
  )
}

Field.defaultProps = {
  error: false,
  hint: null,
  description: null,
  customDescription: null,
  disabled: false
}

Field.propTypes = {
  /** Id for Input and Label*/
  id: PT.string.isRequired,
  /** Label */
  label: PT.string.isRequired,
  /** Disabled input children */
  disabled: PT.bool,
  /** Hint */
  hint: PT.string,
  /** Description */
  description: PT.string,
  /** Custom Description (react element to replace default description)*/
  customDescription: PT.element,
  /** Error variant */
  error: PT.bool,
  /** Childen input node */
  children: PT.oneOfType([
    PT.arrayOf(PT.node),
    PT.node
  ]).isRequired
}

export { Field }
