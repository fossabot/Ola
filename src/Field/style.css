.field {

}
.field-label {
    color: var(--black);
    display: flex;
    justify-content: space-between;
    font: var(--font-callout);
    font-weight: var(--bold);
}
.field-hint {
    font: var(--font-caption);
    margin-left: 1em;
    font-weight: var(--regular);
    color: var(--gray);
}
.field-input {
    margin: var(--size-2) 0;
}
.field-description,
.field-error {
    font: var(--font-caption);
    margin-top: 0;
    color: var(--gray);
}
.field-error {
    color: var(--error);
}
.field.is-invalid {
    & .field-label {
        color: var(--error);
    }
}
