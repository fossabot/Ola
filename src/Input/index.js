import React from 'react'
import {default as PT} from 'prop-types'
import cx from 'classnames'

const Input = ({ className, error, type, handleInputValue, maxCharacter, ...props }) => {
  
  const handleChange = (event) => {
    handleInputValue(event.target.value.length)
  }
  return type === 'textarea' ?
    ( <textarea type={type} className={ cx('ola_input', {'is-invalid': error}, className) } {...props} /> ) :
    ( <input maxLength={maxCharacter} type={type} onChange={handleChange}  className={ cx('ola_input', {'is-invalid': error}, className) } {...props} /> )
}

Input.defaultProps = {
  type: 'text',
  className: null,
  error: false
}

Input.propTypes = {
  /** Input valid types (texarea return texarea html tag) */
  type: PT.oneOf([
    'date',
    'datetime-local',
    'email',
    'month',
    'number',
    'password',
    'search',
    'tel',
    'text',
    'textarea',
    'time',
    'url',
    'week'
  ]),
  /** Extra className */
  className: PT.string,
  /** Input is invalid */
  error: PT.bool,
  /** Counter props  */
  handleInputValue: PT.func,
  maxCharacter: PT.number
}

export default Input
