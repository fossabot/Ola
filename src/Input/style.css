.input {
    --border: var(--gray-light);
    --color: var(--black);
    --background: var(--white);
    --focus: var(--brand-focus);

    display: block;
    box-sizing: border-box;
    height: var(--size-8);
    width: 100%;
    border: solid 2px var(--border);
    background: var(--white);
    color: var(--color);
    background-color: var(--background);
    border-radius: var(--radius-small);
    font: var(--font-callout);
    padding: 0 1em;
    transition-property: background-color, color, box-shadow, border-color;
    transition-duration: 200ms;
    outline: 0;
    white-space: nowrap;
    text-decoration: none;
    font-weight: var(--regular);

    &::placeholder {
        color: var(--gray-light);
        opacity: 1;
    }

    &:focus {
        --border: var(--brand);
        box-shadow: 0 0 0 4px var(--focus);
    }

    &:invalid,
    &.is-invalid {
        --border: var(--error);
        --color: var(--error);
        box-shadow: none;

        &:focus {
            box-shadow: 0 0 0 4px var(--error-focus);
        }
    }

    &:disabled {
        --border: var(--gray-xlight);
        --color: var(--gray-light);
        cursor: not-allowed;

        &::placeholder {
            opacity: 0;
        }
    }
}
