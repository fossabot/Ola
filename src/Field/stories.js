import React, { useState } from 'react'
import Field from './'
import Input from '../Input'
import Button from '../Button'

import { useForm, Controller } from 'react-hook-form'

export default {
  title: 'Field',
  component: Field,
  args: {
    id: 'field-test',
    label: 'Label of the field',
    description: 'This is a description',
    hint: '(optional)',
  },
}

export const Base = (args) => (
  <Field {...args}>
    <Input placeholder="Text placeholder" />
  </Field>
)

export const reactHookFormsExample = () => {
  const onSubmit = (data) => setData(data)
  const {
    handleSubmit,
    control,
    formState: { errors },
  } = useForm()

  const [data, setData] = useState(null)

  return (
    <>
      <form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          name={'testInput'}
          control={control}
          defaultValue={'Default value'}
          rules={{ required: 'illo k ase loko?' }}
          render={({ field }) => (
            <Field
              id="field-test"
              label={'Test input'}
              error={!!errors?.testInput?.message}
              description={
                errors?.testInput?.message ? errors.testInput.message : ''
              }
            >
              <Input {...field} />
            </Field>
          )}
        />
        <Button variant="primary" type="submit">
          Submit
        </Button>
      </form>
      {data && <code>{JSON.stringify(data)}</code>}
    </>
  )
}

Base.argTypes = {
  children: { control: { disable: true } },
}
