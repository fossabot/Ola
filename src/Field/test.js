import React from 'react'
import Field from './'
import Input from '../Input'
import renderer from 'react-test-renderer'

it('Default Field', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label"><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with description', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" description="Test Description"><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with custom description', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" description={<span>Custom description</span>}><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with hint', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" hint="Test hint"><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with error', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" error={true}><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with custom description with error', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" error={true} description={<span>Custom description</span>}><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Field with character counter', () => {
  const tree = renderer
    .create(<Field id="test" label="Test label" error={true} maxCharacter={true}><Input /></Field>)
    .toJSON()
  expect(tree).toMatchSnapshot()
})