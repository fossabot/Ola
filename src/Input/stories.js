import React from 'react'
import Input from './'

export default {
  title: 'Input',
  component: Input,
  args: {
    disabled: false,
    readOnly: false
  }
}

export const Base = (args) => <Input {...args}/>
