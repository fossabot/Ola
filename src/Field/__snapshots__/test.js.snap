// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default Field 1`] = `
<div
  className="ola_field"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;

exports[`Field with character counter 1`] = `
<div
  className="ola_field is-invalid"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;

exports[`Field with custom description 1`] = `
<div
  className="ola_field"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
  <p
    className="ola_field-description"
  >
    <span>
      Custom description
    </span>
  </p>
</div>
`;

exports[`Field with custom description with error 1`] = `
<div
  className="ola_field is-invalid"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
  <p
    className="ola_field-error"
  >
    <span>
      Custom description
    </span>
  </p>
</div>
`;

exports[`Field with description 1`] = `
<div
  className="ola_field"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
  <p
    className="ola_field-description"
  >
    Test Description
  </p>
</div>
`;

exports[`Field with error 1`] = `
<div
  className="ola_field is-invalid"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;

exports[`Field with hint 1`] = `
<div
  className="ola_field"
>
  <label
    className="ola_field-label"
    htmlFor="test"
  >
    Test label
    <span
      className="ola_field-hint"
    >
      Test hint
       
      <strong />
    </span>
  </label>
  <div
    className="ola_field-input"
  >
    <input
      className="ola_input"
      disabled={false}
      id="test"
      maxLength={null}
      onChange={[Function]}
      type="text"
    />
  </div>
</div>
`;
